marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs

```diff
@@ -35,7 +84,137 @@ buildApplyInputsConstraints
+  -- Consume UTXOs containing Marlowe script.
+  tell $ mustConsumeMarloweOutput invalidBefore' invalidHereafter' redeemer
+
+  -- Consume UTXOs containing all necessary role tokens and send them back.
+  for_ requiredParties $ traverse_ $ \case
+    V1.Role role -> tell $ mustSpendRoleToken $ roleAssetId role
+    _ -> pure ()
+
+  -- Require signature of an every party which is autorhized through an address.
+  for_ requiredParties $ traverse_ $ \case
+    V1.Address _ address -> case address of
+      P.Address (P.PubKeyCredential (P.PubKeyHash pkh)) _ -> do
+        tell $ requiresSignature $ PaymentKeyHash $ P.fromBuiltin pkh
+      _ -> pure ()
+    _ -> pure ()
+
+  -- Apply inputs.
+  let
+    txInterval = (invalidBefore', invalidHereafter')
+    transactionInput = V1.TransactionInput { txInterval, txInputs = redeemer }
+  (possibleContinuation, payments) <- case V1.computeTransaction transactionInput state contract of
+     V1.Error err -> throwError $ ApplyInputsConstraintsBuildupFailed (MarloweComputeTransactionFailed $ show err)
+     V1.TransactionOutput _ payments _ V1.Close ->
+       pure (Nothing, payments)
+     V1.TransactionOutput _ payments state' contract' ->
+       pure (Just (state', contract'), payments)
+
+  -- Construct outputs constraints.
+  -- Require Marlowe output if the contract is not closed.
+  case possibleContinuation of
+    Just (state'@V1.State { accounts }, contract') -> do
+      let
+        datum' = V1.MarloweData params state' contract'
+        assets = moneyToAssets $ V1.totalBalance accounts
+      tell $ mustSendMarloweOutput assets datum'
+    Nothing -> pure ()
+
+  -- For every payment require an output either to the role
+  -- payout script or to directly to the party address.
+  for payments \(V1.Payment _ payee token quantity) -> do
+    let
+      assets = case token of
+        V1.Token "" "" -> Assets (Lovelace $ fromInteger quantity) mempty
+        V1.Token cs tn -> do
+          let
+            assetId = toAssetId cs tn
+            quantity' = CS.Quantity $ fromInteger quantity
+          Assets (Lovelace 0) (CS.Tokens $ Map.singleton assetId quantity')
+    case payee of
+      V1.Party (V1.Address net addr) -> do
+        let addr' = Address $ V1.serialiseAddress net addr
+        tell $ mustPayToAddress assets addr'
+      V1.Party (V1.Role role) -> do
+        tell $ mustSendPayoutOutput assets $ roleAssetId role
+      V1.Account _ -> pure ()
+
+  where
+    utcToSlotNo :: SlotConfig -> UTCTime -> SlotNo
+    utcToSlotNo = error "not implemented"
+
+    slotStart :: SlotConfig -> SlotNo -> P.POSIXTime
+    slotStart = error "not implemented"
+
+    marloweInputContent (V1.NormalInput c) = c
+    marloweInputContent (V1.MerkleizedInput c _ _) = c
+
+    marloweInputParty = marloweInputContent >>> \case
+      V1.IDeposit _ party _ _         -> Just party
+      V1.IChoice (V1.ChoiceId _ party) _ -> Just party
+      V1.INotify                      -> Nothing
+
```

Comment:
@jhbertra The below pieces are literal copies of the code from the test suite. Do you think that we should extract these conversions into a separate place? Any suggestions where to put them in such a case?

Reply:
Yes, I do have a suggestion, though I would do this as a dedicated PR. We already have a `Language.Marlowe.Runtime.Cardano.Api` module which defines all the `Cardano.Api <-> Language.Marlowe.Runtime.ChainSync.Api`` conversions. Let's make a `Language.Marlowe.Runtime.Plutus.Api` module that does the same thing with `Plutus` types.